### PR TITLE
Edge Masking to generate Explanations

### PR DESCRIPTION
**Pull request by fork123aniket**
This PR is an extension of discussion #4440. Moreover, this PR contains implementation of how to compute layer-wise weights for each edge in order to produce explanations for both node-level and graph-level tasks. Furthermore, this implementation is different from authors' original implementation and is fast and more memory efficient than theirs.

### CODE REVIEW

**Comment by Padarn** [comment]
Should we add some tests? I'm curious why codecov is not showing coverage checks on this?

Also do you want to add an import to `torch_geometric/nn/models/__init__.py`?

File: torch_geometric/nn/models/GraphMask.py [hunk 1]
```diff
@@ -0,0 +1,840 @@
+import math
+from inspect import signature
+from math import sqrt
+from typing import Optional
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from originalMessagePassing import MessagePassing
```
Comment (Padarn) [inline torch_geometric/nn/models/GraphMask.py:9]: is this a mistake? I can't find this module

File: torch_geometric/nn/models/GraphMask.py [hunk 2]
```diff
@@ -0,0 +1,840 @@
+import math
```
Comment (Padarn) [inline torch_geometric/nn/models/GraphMask.py:1]: I think the filename should be `graph_mask.py`

File: torch_geometric/nn/models/GraphMask.py [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+            penalty = torch.zeros_like(input_element)
+
+        if summarize_penalty:
+            penalty = penalty.mean()
+
+        s = s * (zeta - gamma) + gamma
+
+        clipped_s = s.clamp(min_val, max_val)
+
+        if True:
```
Comment (Padarn) [inline torch_geometric/nn/models/GraphMask.py:161]: mistake?

File: torch_geometric/nn/models/GraphMask.py [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+                        pbar.set_description(f'Explain node {node_idx}')
+                output = self.full_biases[i]
+                for j in range(len(gate_input)):
+                    partial = self.gates[i * 4][j](gate_input[j][i])
+                    result = self.gates[(i * 4) + 1][j](partial)
+                    output = output + result
+                relu_output = self.gates[(i * 4) + 2](output / len(gate_input))
+                sampling_weights = self.gates[(i * 4) +
+                                              3](relu_output).squeeze(dim=-1)
+                sampling_weights, penalty = self.__hard_concrete__(
```
Comment (Padarn) [inline torch_geometric/nn/models/GraphMask.py:505]: penalty is unused here btw (not sure if its intentional)

**Comment by Padarn** [comment]
Reading this I was very confused at first, I got the context from the discussion linked though, thanks!

It makes me wonder however whether we should separate the functionality here, perhaps we could create a 

`ExplainableMessagePassing` class that looking something like 
```
class ExplainableMessagePassing(MessagePassing):
    def explain_message(self, out, x_i, x_j): 
        ...
   def from_message_passing(self, module: MessagePassing):
       ...
```
where `from_message_passing` would do all of he injection work which is currently inside the GraphMask model?

**Comment by fork123aniket** [comment]
Closing this PR. Will upload a new PR containing correct file.
